server/api/attendance/attendance.controller.js

```diff
@@ -0,0 +1,674 @@
+        var classYearId = classYear._id;
+        getAttendance(userId, classYearId, function(userAttendance){
+            return res.json(userAttendance);
+        });
+    });
+};
+exports.getAttendanceMe = function(req, res) {
+    req.params.user = req.user._id
+    exports.getAttendance(req,res);
+};
+// *******************************************************
+
+
+// *******************************************************
+// Get attendance for a specific user (or current user) on a date
+//
+// router.get('/present/:user/today', auth.hasRole('mentor'), controller.present);
+// router.get('/present/:user/:date', auth.hasRole('mentor'), controller.present);
+//
+// router.get('/present/me/today', auth.isAuthenticated(), controller.presentMe);
+// router.get('/present/me/:date', auth.isAuthenticated(), controller.presentMe);
+
+// Check if a userId is present today
+var getPresent = function(userId, date, classYearId, cb){
+    var callback = cb || function(){};
+    Attendance.find({user:userId, date: date, classYear:classYearId}, function (err, attendance) {
+        return callback(attendance);
+    });
+};
+
+exports.present = function(req, res) {
+    var date = req.params.date;
+    if (req.params.date === 'today'){
+        date = isoDateToTime(new Date());
+    }
+    else{
+        date = isoDateToTime(req.params.date);
+    }
+    var userId = req.params.user;
+    return ClassYear.getCurrent(function(err, classYear){
+        var classYearId = classYear._id;
+        getPresent(userId, date, classYearId, function(userAttendance){
+            return res.json(userAttendance);
+        });
+    });
+};
+exports.presentMe = function(req, res) {
+    req.params.user = req.user._id
+    exports.present(req,res);
+};
+// *******************************************************
+
+// *******************************************************
+// Mark attendance as present, subject to verification
+// router.post('/attend', auth.isAuthenticated(), controller.attend);
+exports.attend = function(req,res){
+    var user = req.user;
+    var code = req.body.dayCode;
+    if (!code) {return res.send(400, 'No Code Submitted');}
+    // Uppercase code from client so it is case-insensitive. This must happen
+    // after the above check, otherwise toUpperCase() might not exist.
+    code = code.toUpperCase();
+    // Check code against current class year
+    return ClassYear.getCurrent(function(err, classYear){
+        if (err) {return res.send(500, err);}
+        return getPresent(user._id, isoDateToTime(new Date()), classYear._id, function(userAttendance){
+            var submittedFull = false;
+            var submittedSmall = false;
+            var submittedSmallBonus = false;
+            var submittedFullBonus = false;
+            // Check what types of attendance the user has submitted today
+            for (var a = 0; a < userAttendance.length ; a++){
+                if (userAttendance[a].bonusDay && !userAttendance[a].smallgroup){
```

Comment:
this should be simpllifed with a function that takes the attendance and returns a type. Also shouldn't we only have one attendance item per day?